### ISSUE TITLE
Use LoadLibraryEx instead of SafeLoadLibrary for accessing Python-DLL

### ISSUE THREAD

**Issue by jr2804** [pyscripter/python4delphi#289]
(This issue only affects Windows and maybe only Delphi)

A standard distribution of Python brings their own MSVC redistributables (like e.g., _concrt140.dll, msvcp140.dll, vccorlib140.dll, vcomp140.dll_, etc.), which are exactly matching to python.exe and python 3x.dll and only these are intended to be used.

However, in case you use python4delphi in conjunction with such a distribution, these redistributables cannot be found because they are not in the search path. If no other suitable VC-Redist-Package is found in the search path, you will get a crash. But even worse, if a **different** (and not suitable!) VC-Redist-Package is found somewhere else in the search path, python3x.dll will use these instead of the deployed ones.

It is suggested to resolve this by slightly modifying `procedure TDynamicDll.DoOpenDll(const aDllName : string);` and change
`FDLLHandle := SafeLoadLibrary(GetDllPath+DllName);`
to
`FDLLHandle := LoadLibraryEx(GetDllPath+DllName, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR or LOAD_LIBRARY_DEFAULT_DIRS);`.

This function allows to let the Python-DLL search for dependencies next to its file location. I have tested this here locally and works like a charm. The possible change of FPU control word by the library can be handled as follows (if this is necessary / applicable at all for the python3x.dll??):

`//Store the fpu control word as some dlls might corrupt it when they are not`
`//loaded properly`
`fpu_word := get8087cw;`

`{$IFDEF WIN32}`
`  //On Windows prevent the PE-Loader from showing up any exceptions`
`  SetErrorMode(SEM_FAILCRITICALERRORS);`
 ` {$ENDIF}`

 ` try`
`    result := LoadLibrary(PChar(AModule));`
 ` finally`
`    {$IFDEF WIN32}`
`    //Reset the error mode to default`
`    SetErrorMode(0);`
`    {$ENDIF}`
`    //Reset the FPU word`
`    set8087cw(fpu_word);`
`  end;`
(Credits: [delphipraxis.net](https://www.delphipraxis.net/126308-abfangen-von-exceptions-bei-loadlibrary.html), unfortunately in German)

Regarding `SafeLoadLibrary()`, see also:
[https://blog.dummzeuch.de/2017/06/13/do-not-call-safeloadlibrarydllname-dll/](https://blog.dummzeuch.de/2017/06/13/do-not-call-safeloadlibrarydllname-dll/)

**Comment by pyscripter** [comment]
- The behavior of SafeLoadLibrary  with SEM_NOOPENFILEERRORBOX (default) is exactly what we want.
- The FPU flags need to be managed separately (see https://github.com/pyscripter/python4delphi/wiki/MaskFPUExceptions).
- Regarding the MSVC dependencies see https://github.com/pyscripter/python4delphi/wiki/P4DPython26.  But this is no longer a issue.

I believe in "if it works don't fix it".   What is the problem you are trying to solve?  In my experience there is no problem with loading failing because the dependencies of pythonxy.dll were not found.  Conda distributions come with a number of dlls located alongside pythonxy.dll, but they load fine.   (The problem with conda distributions is that they also load dlls from the \Library\bin subdirectory as discussed https://github.com/pyscripter/python4delphi/wiki/FindingPython, but that would be resolved by your suggestion).

**Comment by jr2804** [comment]
Thanks for the quick reply!

> What is the problem you are trying to solve?

Yes, maybe I should have explained this in a better way. We are not referring to the function `TPythonEngine.SetPythonHome()`, which has an impact after the actual pythonxy.dll was loaded. We had the following issue in the past:

1) Our application embeds a (quite comprehensive) Python distribution, which was based on WinPython until they dropped 32bit support. Then we moved to the basic installer from python.org and extended it with several packages.

2) We wanted to keep the deployment of the distribution clean, since it is not unlikely that another Python installation is on the target PC. That means "our" Python distribution is an _unregistered version_ and is neither added to the search path nor to the registry, but our application knows the exact location.

3) Unfortunately, we had exactly an issue, which you objected:

> "In my experience there is no problem with loading of the python dll failing because the dependencies of pythonxy.dll were not found."

4) Let's assume the following setup: 
a) our app is located at _c:\myapp\app.exe_ and the Python library to embed at _c:\pyredist\python37.dll_.  _c:\pyredist\python.exe_ is present as well, but only used for debug/testing.
b) _python37.dll_ depends on _vcruntime140.dll_, which is provided by the basic python.org installer in the same directory, next to _python37.dll_  (and _python.exe_).
c) Let's assume we are deploying the app and Python distribution on a machine#1, which has VC-Redist located and available somewhere in the search path and and on a machine#2, which does not have any VC-Redist at all.
d) Before loading the DLL in _app.exe_, the current working path may be _c:\myapp\\_ or something else - but not _c:\pyredist\\_.
e) When loading _python37.dll_, it looks for its dependency _vcruntime140.dll_ according to this [search order](https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order).
f) On machine#1, _vcruntime140.dll_ is located via search path, but not in the intended location _c:\pyredist\\_. This *may* work, in case both versions of _vcruntime140.dll_ are compatible. We had the issue in some cases, where they were not. We fixed this with the suggested solution.
g) On machine#2 _vcruntime140.dll_ is only located next to _python37.dll_. I must admit that I cannot remember exactly the behaviour. but either it is not found all (because it is not in the working directory, next to the executable and not in the path; the directory that contains the DLL is not regarded, at least when not using `LoadLibraryEx()` with flag `LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR`) **or** it is finally found next to the *python37.dll*, but definitely after searching in the path.

Reading the [link](https://docs.microsoft.com/en-us/windows/desktop/api/Winbase/nf-winbase-setdlldirectorya) with the DLL search order again, also the usage of[ WinAPI function `SetDllDirectory()`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectoryw) may work here even better ...?
